Added HTTP handler to expose bookie state

merlimat opened the pull request:
### Motivation

Allow to check the state of the bookie through HTTP.

The main reason is to allow light-weight probes that can be run frequently to establish whether the bookie is running and meets a minimum readiness criteria. One example is to automatically wait until the bookie has completed the startup sequence.

### Changes

Added new HTTP handler that exposes the values from the bookie `StateManager`.

eolivelli commented:
@jvrao fyi

merlimat commented:
run integration tests

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/server/http/service/BookieStateService.java:
```diff
@@ -0,0 +1,94 @@
+
+        if (HttpServer.Method.GET != request.getMethod()) {
+            response.setCode(HttpServer.StatusCode.NOT_FOUND);
+            response.setBody("Only support GET method to retrieve GC details."
```
the content in the body doesn't match the endpoint.

merlimat commented:
Updated based on comments. I also added a separate handler `/api/v1/bookie/is_ready`. The reason is that to use this a readyness probe, systems like K8S expects a simple 200 response or error if the process is not ready. Having the JSON response makes it difficult to integrate with that.

eolivelli commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/server/http/TestHttpService.java:
```diff
@@ -826,4 +829,30 @@ public void testGCDetailsService() throws Exception {
+
+        HttpServiceRequest request1 = new HttpServiceRequest(null, HttpServer.Method.GET, null);
+        HttpServiceResponse response1 = bookieStateServer.handle(request1);
+        assertEquals(HttpServer.StatusCode.OK.getValue(), response1.getStatusCode());
```
Can you add a test case for the not-ready case?

merlimat replied:
Added

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/server/http/service/BookieIsReadyService.java:
```diff
@@ -0,0 +1,60 @@
+            response.setCode(HttpServer.StatusCode.OK);
+        } else {
+            response.setCode(HttpServer.StatusCode.SERVICE_UNAVAILABLE);
+            response.setBody("Bookie is not fully started yet");
```
if a bookie is shutting down, `isRunning` is also false. then the message is not correct. Can we check if the state manager is shutting down or not, and return the proper error message?

merlimat replied:
Fixed

merlimat commented:
@sijie @eolivelli Address the comments. Please take a final look.

merlimat commented:
run bookkeeper-server bookie tests
